File: java/questions/QuickSort.java

public class QuickSort{
	private static void swap(int[] array, int x, int y){
		int temp = array[x];
		array[x] = array[y];
		array[y] = temp;
	}
	private static int partition(int[] array, int first, int last){
		int pivot = array[(first + last)/2];
		int i = first - 1;
		int j = last  + 1;
		while(true){
			do{
				i++;
			}while(array[i] < pivot);
			do{
				j--;
			}while(array[j] > pivot);
			if(i < j)
				swap(array, i, j);
			else
				return j;

		}
	}
	static void qsort(int[] array, int first, int last){
		int split = partition(array, first, last);
		if(first < split)
			qsort(array, first, split);
		if(last > split + 1)
			qsort(array, split+1, last);
	}
	static void quickSort(int[] array){
		qsort(array, 0, array.length - 1);
	}
	public static void main(String[] args){
		int[] array = {10, 5 , 4 , 88, 99, 1, 9, 23, 75, 100, 234, 55};
		quickSort(array);
		for(int i = 0; i < array.length; i++){
			System.out.print(" " + array[i]);
		}
		System.out.println();
	}
}

File: misc/quicksort.c
#include <stdio.h>
#include <stdlib.h>

#define N 12

void swap(int *arr, int a, int b)
{
    int tmp = arr[a];
    arr[a] = arr[b];
    arr[b] = tmp;
}
void
print(int* arr, int left, int right){
    for(int i = left; i <= right; i++){
        printf("%d ", arr[i]);
    }
    printf("\n");    
}
int 
pertition(int* arr, int left, int right){

    int pivot =  arr[(left + right)/2];
    printf("pivot [%d] => %d\n", (left + right)/2, pivot);
    int i = left - 1;
    int j = right + 1;
    while(1){
        do{
            i++;
        }while(arr[i] < pivot);
        do{
            j--;
        }while(arr[j] > pivot);
        if(i < j)
            swap(arr, i, j);
        else{
            print(arr, left, right);
            printf("split --> %d\n",j);
            return j; // arr[j] = end of left array
        }
    }
}
void 
quicksort(int* arr, int left, int right){
    int split = pertition(arr, left, right);
    if(left < split)
        quicksort(arr, left, split); // left sub array
    if(right > split + 1)
        quicksort(arr, split + 1, right); // right sub array
}



int
main(int argc, char** argv)
{
    int arr[N] = {19, 39, 3, 33, 4, 30, 98, 83, 77, 5, 0, 40};
    print(arr, 0, N - 1);
    quicksort(arr, 0, N - 1);
    print(arr, 0, N - 1);
}